### PR TITLE
chore(dist): suppress Kryo illegal access warnings

### LINKED ISSUE #3711: RFE: Suppress llegal reflective access warning
At the moment (0.22.1), the following message appears when starting the broker:

```
zeebe_broker | WARNING: An illegal reflective access operation has occurred
zeebe_broker | WARNING: Illegal reflective access by com.esotericsoftware.kryo.util.UnsafeUtil (file:/usr/local/zeebe/lib/kryo-4.0.2.jar) to constructor java.nio.DirectByteBuffer(long,int,java.lang.Object)
zeebe_broker | WARNING: Please consider reporting this to the maintainers of com.esotericsoftware.kryo.util.UnsafeUtil
zeebe_broker | WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
zeebe_broker | WARNING: All illegal access operations will be denied in a future release
```

Can we suppress this message? According to [this comment](https://github.com/EsotericSoftware/kryo/issues/543#issuecomment-395959003) you can suppress it if you add the following to your JVM args `--add-opens=java.base/java.nio=ALL-UNNAMED`.

Since there is nothing for users to do about this, they don't need to see this in the logs.

### PR DESCRIPTION
**Pull request by npepinpe**
## Description

- opens Java 9 modules for reflection access required by Kryo, suppressing the illegal access warnings at start up

## Related issues

closes #3711

## Pull Request Checklist

- [x] All commit messages match our [commit message guidelines](https://github.com/zeebe-io/zeebe/blob/develop/CONTRIBUTING.md#commit-message-guidelines)
- [x] The submitting code follows our [code style](https://github.com/zeebe-io/zeebe/wiki/Code-Style)
- [x] If submitting code, please run `mvn clean install -DskipTests` locally before committing

### CODE REVIEW

**Comment by Zelldon** [comment]
Actually I hesitate to approve this because I'm not a fan of just hide warnings, such things are then forgotten. Is there the possibility to just fix this issue? 

But this is just a problem if we migrate to the next java version that these API's do not exist anymore right? Then we might have to fix it then and I'm probably ok with hiding them until then.

Would like to hear @menski opinion on this.

**Comment by npepinpe** [comment]
Thanks, which is why I requested a review from you :) I think this case is a little different, let's discuss it f2f first.

**Comment by jwulf** [comment]
I’m not a fan of hiding warnings either, but we should either fix it or suppress it, because we are exposing it to users and there is nothing they can or should do do about it, except open a GitHub issue and ask for it to be fixed or suppressed.

**Comment by npepinpe** [comment]
I originally wanted to discuss it in person as that's probably faster, but I guess it'll be a while before that's possible :sweat_smile: 

So TL;DR of my original thought process, suppressing this warning was going to affect production usage only - every time we ran the broker locally (e.g. tests, IDE, etc.) we would get this warning. Only when ran through the distribution's `bin/broker` would the warning be suppressed. I figured it was the lesser of two evils, as the next Kryo version will not be backwards compatible (:scream:) and as far as I can tell, they aren't working on fixing any of these warnings (as of now anyway).

> Kryo works fine with Java 11. When reflection is disallowed (note it is allowed in Java 11 by default), many Kryo serializers will no longer work. Kryo itself should still work fine, but you would need to use serializers that don't rely on reflection. I don't personally have a need to use Java with reflection disabled, so I don't currently have plans to work on alternate serializers.

So from this comment, we can infer two things:

1. These warnings won't go away with default Kryo usage
1. We can workaround them by using different serializers for classes that we cannot openly access

To get rid of all these pesky warnings, we have to get rid of reflection on Java internal APIs. This means we can't use the `FieldSerializer` to serialize private fields and classes that are not opened to our project. For all of these, we need to either implement custom serializers, or use slower serialization methods (e.g. Java serialization :nauseated_face:). 

So I would propose that we actually move to `--illegal-access=deny` instead of `--ilegal-access=permit`, which means the next time we introduce one of these issues, we wouldn't get a warning but rather it would crash. What I understand is when registering classes with Kryo, the first thing Kryo does (if the class should be serialized with `FieldSerializer`) is build a cache of the class fields through reflection, marking them as accessible if need be. So we should be catching all errors on start up since Atomix pre-registers all its serializable classes with Kryo. I'll run a benchmark anyway to make sure this won't blow up in our faces, but if consensus is having `--illegal-access=deny` too dangerous, I'm fine with keeping it permit (it just means we have to be more disciplined if the warning appears again).

**Comment by npepinpe** [comment]
See https://github.com/zeebe-io/atomix/pull/152 for Atomix side fix

**Comment by npepinpe** [comment]
I ran a benchmark with and without, and didn't see any notable different in throughput or memory usage, nor any warnings or exceptions (even with `--illegal-access=deny`). Throughput seems to decrease overtime but that was also true without this PR so :man_shrugging:

**Comment by npepinpe** [comment]
Seems to be some issue with the go tests, I'll look into it tomorrow - the rest seems to work fine.

**Comment by npepinpe** [comment]
Aaand of course today the Go tests pass without issues :thinking:

**Comment by Zelldon** [comment]
@npepinpe  Do we still need this you changed something in atomix right?

**Comment by npepinpe** [comment]
We already changed and merged it. But it means we have to be more careful what we serialize in Atomix, so we should consider switching to Java 11 and disabling it there as well to catch those issues in Atomix before they arise in Zeebe.

**Comment by Zelldon** [comment]
>  so we should consider switching to Java 11

This is on my todo list yes :smile: 

Can we close this then? Otherwise we would not see this warnings we do something wrong right :sweat_smile:

**Comment by npepinpe** [comment]
I changed the PR since the initial one, it now switches to `--illegal-access=deny`, so we won't see warnings, it will just fail to start.

**Comment by Zelldon** [comment]
Ok got it thanks

**Review by ChrisKujawa — APPROVED** [review]

**Comment by npepinpe** [comment]
First time I see this, the build succeeded but bors marked the merge as failed... :thinking: 

bors retry

**Comment by Zelldon** [comment]
> Cannot contact zeebe-ci-build-staging-1064-t8lt7-3cjkv: hudson.remoting.ChannelClosedException: Channel "unknown": Remote call on JNLP4-connect connection from 10.110.25.6/10.110.25.6:36338 failed. The channel is closing down or has closed down

> missing workspace /home/jenkins/agent/workspace/zeebe-io_zeebe_staging on zeebe-ci-build-staging-1064-t8lt7-3cjkv

I dont think that it was successful :smile:

**Comment by npepinpe** [comment]
The one I retried wasn't, but the one before had a green checkmark (see https://bors.zeebe.camunda.cloud/batches/848). Either way, that looks like k8s killing my pod :(

**Comment by npepinpe** [comment]
bors retry

**Comment by npepinpe** [comment]
bors retry

**Comment by npepinpe** [comment]
bors retry

**Comment by npepinpe** [comment]
bors retry

**Comment by npepinpe** [comment]
bors retry

**Comment by npepinpe** [comment]
How did bors fail before the build even finished?

**Comment by npepinpe** [comment]
Hell the build just succeeded and bors already failed??

**Comment by npepinpe** [comment]
image

**Comment by npepinpe** [comment]
Still no idea why it would fail before the build was even finished (and the build even succeeded after). Will try one last time, and also look into it with infra.

This PR is doomed!

source

**Comment by npepinpe** [comment]
bors retry

(or die trying)

**Comment by Zelldon** [comment]
Haha how much have you paid to bors to merge this 😀

**Comment by npepinpe** [comment]
giphy-downsized

😂 But in all seriousness, it kept passing the branch builds but not with bors? And it was making bors fail but not the check? Never seen this before. I'll have to ask infra to look into the bors logs and see if it crashed or something 👀 

This shouldn't even impact our tests anyway since it only changes the distribution?? 🤷‍♂
